### PR TITLE
Add to Editor tests the ability to close the editor screen when Screen Object fails to load

### PR DESCRIPTION
**Pull request by tiagomar**
One identified UI Tests issue happens when the Editor screen takes longer to load. Some elements would load right away like the ones In Navigation Bar, but others may take some more seconds to load, like the editor blocks and the `add-block-button`. 

image

This issue doesn't reproduce consistently so it would most likely pass in the automatic retry, however the `tearDown()` and `setUp` can't handle the Editor screen, so all the following tests fail. 
The purpose of this PR is to enable Edior tests' `tearDown()` to properly close the Editor when the Editor Screen Object fails initializing.

* Changes in `tearDown()`
When the issue happens `editorScreen()` evaluates to `nil`, a check was added and `closeEditorDiscardChanges()` will run in that case .

* Changes in `BlockEditorScreen`
  * `expectedElementGetters`
  The order of the elements was changed so the element that loads first is the first to be checked.
  * Increased timeout 
  The initial idea was not to increase the timeout for time sake, however from the evidences we have, the button did load right before the 7 secs timeout but it was not enough time to check for the `Close` button that was already there right from the start. Also adding 3 more seconds to the execution, in the worst scenario, is still much better than rerunning the test in the automatic retry.

To test:
Running locally:
In BlockEditorScreen, change the add block button name from `add-block-button` to something else.

Running on CI:
I don't know how to reproduce it in CI but I have re-enabled the Editor tests for testing this PR. If you we are lucky the issue will happen and we will still see all the other tests running normally, including the failed test retries.

**P.S.** I'll disable the Editor tests before merging this PR.

## Regression Notes
1. Potential unintended areas of impact
Editor UI Tests

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Ran the tests in CI

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:24]: > The order of the elements was changed so the element that loads first is the first to be checked.

Great idea!

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:26]: > [...] adding 3 more seconds to the execution, in the worst scenario, is still much better than rerunning the test in the automatic retry.

Good point. 👍

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: Have you thought about making this a `static` method on `EditorScreen`?

Defined this way, the function is a _free function_, a function that's not attached to any type. The Swift API guidelines [recommend against](https://www.swift.org/documentation/api-design-guidelines/#prefer-method-and-properties-to-functions) defining them, unless in special cases.

While there's no explicit rationale there, my point of view is that free functions belong to the global namespace, that is, they are available everywhere.

The situation wouldn't be much different if this was a `static` method on `EditorScreen`, but at least if would be clearer where it belongs.

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: Swift tip: The word "generic" has a different meaning in the context of Swift. See the official documentation for [_Generics_](https://docs.swift.org/swift-book/LanguageGuide/Generics.html.

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: Nitpick: If you want to make code stand out in comments, backticks are more appropriate.

```suggestion
// making the Editor Screen evaluate to `nil`
```

I say this because backticks parse as inline code in Markdown.

In fact, if we were to make that comment a documentation comment, Xcode would render them like this:

image

vs with the single quotes:

image

File: WordPress/WordPressUITests/Tests/EditorGutenbergTests.swift
Comment (mokagio) [inline WordPress/WordPressUITests/Tests/EditorGutenbergTests.swift]: I haven't tried running it, but I think we could combine this `if` with the one above:

```swift
if editorScreen == nil {
  closeEditorDiscardChanges()
} else if TabNavComponent.isVisible == false() { /* if the code hits this else, editorScreen must not be nil, no need to check it */
  // ...
}
```

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: What do you think of the following name, to make it clearer that we are closing _and_ discarding?

```suggestion
public func closeEditorDiscardingChanges() {
```

Or, alternatively, this:

```suggestion
public func closeEditorAndDiscardChanges() {
```

(_Of the two, my favorite is the first_)

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: Another little nitpick. It's often recommended to keep assignments as close as possible to the code that uses them (spacial locality).

```suggestion
    let closeButton = app.buttons["Close"]
    if closeButton.waitForIsHittable() { closeButton.tap() }
    
    let discardChangesButton = app.buttons["Discard"]
    if discardChangesButton.waitForIsHittable() { discardChangesButton.tap() }
```

The advantage of this style is that one doesn't need to look far to know where `closeButton` or `discardChangesButton` come from when seeing them used in their respective `if` statement.

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [inline WordPress/UITestsFoundation/Globals.swift]: Nitpick: I would be great to inject the app value as a default parameter.

```suggestion
public func closeEditorDiscardChanges(app: XCUIApplication = XCUIApplication()) {
```

There is no _real_ advantage here, because we don't have an `XCUIApplication` instance to pass to the `closeEditorDiscardChanges` call, but IMHO it's a good practice for an object like the `XCUIApplication` instance, which can be easily shared through the tests, as opposed to initialized every time, because the app the tests run on is always the same.

**Comment by mokagio** [comment]
I cancelled the [original build](https://buildkite.com/automattic/wordpress-ios/builds/4992#_) for [a2abd3a](https://github.com/wordpress-mobile/WordPress-iOS/pull/17903/commits/a2abd3a87e846b4ba1637a834120e676620eacb9) because the UI tests on iPhone got stuck:

image

This gives us a chance to re-run the UI tests on iPad, too, to see if they pass again.

**Review by mokagio — COMMENTED** [review]
Overall great approach and detective work to come up with it. 

Please don't take my various comments as a negative statement on the work, I think it's really well done and I'm ever so thankful you're working on this 🙇‍♂️ They are merely suggestions to make the code more _Swifty_.

> Running on CI:
>
> I don't know how to reproduce it in CI but I have re-enabled the Editor tests for testing this PR. If you we are lucky the issue will happen and we will still see all the other tests running normally, including the failed test retries.
>
> P.S. I'll disable the Editor tests before merging this PR.

Something I find useful to both test weird stuff on CI and also remove the need to remember to undo temporary changes is to create a demo PR from the one you want to test with all the necessary hacks. See this one, for example: https://github.com/wordpress-mobile/WordPress-iOS/pull/17746.

Here's two PRs that:

- Replicate your suggestion to ensure the add button is never found https://github.com/wordpress-mobile/WordPress-iOS/pull/17906
- Run the tests 5x times to see if they all pass: https://github.com/wordpress-mobile/WordPress-iOS/pull/17907

File: WordPress/UITestsFoundation/Globals.swift
Comment (tiagomar) [reply WordPress/UITestsFoundation/Globals.swift]: > Have you thought about making this a static method on EditorScreen?

I haven't really, but it makes total sense to me. I'll do it. =)

File: WordPress/UITestsFoundation/Globals.swift
Comment (tiagomar) [reply WordPress/UITestsFoundation/Globals.swift]: TIL

**Comment by tiagomar** [comment]
Hey @mokagio! Thanks for the review! I'm always up for _Swifting_ my code. =D

**Comment by jkmassel** [comment]
You can test the <strong>Jetpack</strong> changes on this Pull Request by downloading it from AppCenter <a href='https://install.appcenter.ms/orgs/automattic/apps/jetpack-installable-builds/'>here</a> with build number: <code>pr17903-3086c2a</code>. IPA is available <a href='https://appcenter-filemanagement-distrib4ede6f06e.azureedge.net/7f71e0d7-ad0f-49f5-b50b-3ed4ab70f7ac/Jetpack%20Alpha.ipa?sv=2019-02-02&sr=c&sig=3VAwLyH6jXeTtz50GVYprxSaKdnbdomEoKM9ESGmToU%3D&se=2022-02-09T15%3A25%3A57Z&sp=r'>here</a>. If you need access to this, you can ask a maintainer to add you.

**Comment by jkmassel** [comment]
You can test the <strong>WordPress</strong> changes on this Pull Request by downloading it from AppCenter <a href='https://install.appcenter.ms/orgs/automattic/apps/WPiOS-One-Offs/'>here</a> with build number: <code>pr17903-3086c2a</code>. IPA is available <a href='https://appcenter-filemanagement-distrib1ede6f06e.azureedge.net/fe526c7b-8e6b-44fa-b9d0-b411ddc439ff/WordPress%20Alpha.ipa?sv=2019-02-02&sr=c&sig=kq21sulUiQAm4FQARDFqbiizLU6IaSRCHJIWUG%2BTMa8%3D&se=2022-02-09T15%3A27%3A43Z&sp=r'>here</a>. If you need access to this, you can ask a maintainer to add you.

**Comment by tiagomar** [comment]
Hey @mokagio , do you think there's anything still blocking this PR from being merged?

**Review by pachlava — APPROVED** [review]
Great work @tiagomar ! It looks like there already were a bunch of improvements while you worked together with @mokagio on this PR. I have nothing else to add to this.

The only thought I had is refreshing (instantiating) the `editorScreen` in teardown before the `if` block, but it doesn't seem necessary, because the execution will jump right to `tearDown` in case if initial `editorScreen` was `nil`.

Thanks!

I think once we finally have three PRs with UI tests stability improvements merged, it's time to try enabling all tests again and see how they work on a large scale.

**Comment by mokagio** [comment]
@tiagomar sorry for letting this sit here 🤦‍♂️  I had missed your ping, then went AFK 😳 Thanks @pachlava for unblocking this 🙌

**Pull request merged**
